pkg/backends/calico/routes.go

```diff
@@ -14,31 +14,265 @@
-	routeString := os.Getenv("CALICO_STATIC_ROUTES")
+	ch := make(chan struct{})
+	go rg.svcInformer.Run(ch)
+	go rg.epInformer.Run(ch)
+
+	/*
+		routeString := os.Getenv("CALICO_STATIC_ROUTES")
+		cidrs := []string{}
+		for _, route := range strings.Split(routeString, ",") {
+			cidr := strings.TrimSpace(route)
+			if cidr != "" {
+				cidrs = append(cidrs, cidr)
+			}
+		}
+		rg.client.updateRoutes(cidrs)
+	*/
+	return
+}
+
+// hash is namespace.name
+func hash(meta *metav1.ObjectMeta) string {
+	return fmt.Sprintf("%s.%s", meta.Namespace, meta.Name)
+}
+
+// onSvcAdd is called when a k8s service is created
+// - creates a serviceEndpoints object and places it in the map
+func (rg *routeGenerator) onSvcAdd(obj interface{}) {
+	log.Info("+onSvcAdd")
+	rg.Lock()
+	defer rg.Unlock()
+
+	svc := obj.(*v1.Service)
+	key := hash(&svc.ObjectMeta)
+	log.WithFields(log.Fields{
+		"key":       key,
+		"serviceIP": svc.Spec.ClusterIP,
+	}).Info("+onSvcAdd")
+	rg.svcEpMap[key] = serviceEndpoints{
+		serviceIP:   svc.Spec.ClusterIP,
+		endpointIPs: make(map[string]bool),
+	}
+
+	// no need to call updateRoutes since the serviceEndpointIPs are empty
+}
+
+// onSvcUpdate is called when a k8s service is updated
+// - clusterIP is an immutable field
+// - only change this method will handle is name/space change
+func (rg *routeGenerator) onSvcUpdate(oldObj, newObj interface{}) {
+	log.Info("+onSvcUpdate")
+	oldSvc, newSvc := oldObj.(*v1.Service), newObj.(*v1.Service)
+
+	// pass if the name/space fields are the same
+	if oldSvc.ObjectMeta.Namespace == newSvc.ObjectMeta.Namespace && oldSvc.ObjectMeta.Name == newSvc.ObjectMeta.Name {
+		return
+	}
+
+	rg.Lock()
+	defer rg.Unlock()
+
+	// update the map
+	oldSvcKey := hash(&oldSvc.ObjectMeta)
+	rg.svcEpMap[hash(&newSvc.ObjectMeta)] = rg.svcEpMap[oldSvcKey]
+	delete(rg.svcEpMap, oldSvcKey)
+
+	// no need to call updateRoutes since no ips actually changes
+}
+
+// onSvcUpdate is called when a k8s service is deleted
+// - removes the serviceEndpoint object from the map
+func (rg *routeGenerator) onSvcDelete(obj interface{}) {
+	log.Info("+onSvcDelete")
+	rg.Lock()
+	defer rg.Unlock()
+
+	svc := obj.(*v1.Service)
+	delete(rg.svcEpMap, hash(&svc.ObjectMeta))
+	rg.updateRoutes()
+}
+
+// waitUntilExists continuously sleeps until the object is found in the svcEpMap
+func (rg *routeGenerator) waitUntilExists(key string) {
+	for {
+		if _, exists := rg.svcEpMap[key]; exists {
+			break
+		}
+		rg.Unlock()
+		time.Sleep(time.Second)
+		rg.Lock()
+		continue
+	}
+}
+
+// onEPAdd is called when a k8s endpoint is created
+// - adds the endpointIPs to the corresponding serviceEndpoints object in the map
+// - sleeps for one second if the serviceEndpoint object does not yet exist
+func (rg *routeGenerator) onEPAdd(obj interface{}) {
+	log.Info("+onEPAdd")
+	ep := obj.(*v1.Endpoints)
+	rg.Lock()
+	defer rg.Unlock()
+
+	// wait until the serviceEndpoint obj exists
+	svcEpKey := hash(&ep.ObjectMeta)
+	rg.waitUntilExists(svcEpKey)
```

Comment:
Oh OK, I guess this solves the problem I mentioned earlier.  At least it will if the two informers are delivering on separate goroutines.  Is that the case?

Reply:
yes, but we're gonna change up the logic a bit